qutip/core/cy/qobjevo.pyx

```diff
@@ -0,0 +1,1114 @@
+                                 "have the signature "
+                                 "`f(t: double, args: dict) -> Qobj`")
+            self.dims = qobj.dims
+            self.shape = qobj.shape
+            self.funcQEvo = set([Q_object])
+            self.elements.append(_FuncElement(Q_object, qobj, args))
+
+        else:
+            raise TypeError("Format not understood")
+
+    def __call__(self, t, args={}):
+        try:
+            t = float(t)
+        except Exception as e:
+            raise TypeError("t should be a real scalar.") from e
+
+        if args:
+            if not isinstance(args, dict):
+                raise TypeError("The new args must be in a dict")
+            old_args = self.args.copy()
+            self.arguments(args)
+            op_t = self._call(t)
+            self.arguments(old_args)
+        else:
+            op_t = self._call(t)
+
+        return Qobj(op_t, dims=self.dims, copy=False)
+
+    cpdef Data _call(QobjEvo self, double t):
+        t += self._shift_dt
+        for func in self.funcQEvo:
+            self.func_qobj[func] = None
+        cdef Data out
+        cdef _BaseElement part = self.elements[0]
+        part.call(t, self.func_qobj)
+        out = _data.mul(part.data, part.coeff)
+        for element in self.elements[1:]:
+            part = <_BaseElement> element
+            part.call(t, self.func_qobj)
+            out = _data.add(out, part.data, part.coeff)
+        return out
+
+    def copy(QobjEvo self):
+        """ Return a copy of this `QobjEvo` """
+        cdef QobjEvo new = QobjEvo()
+        new.args = self.args
+        new.dims = self.dims
+        new.shape = self.shape
+        new._shift_dt = self._shift_dt
+        new.elements = [element.copy() for element in self.elements]
+        new.funcQEvo = self.funcQEvo.copy()
+        new.func_qobj = {}
+        return new
+
+    def arguments(QobjEvo self, new_args):
+        """ Update the arguments """
+        if not isinstance(new_args, dict):
+            raise TypeError("The new args must be in a dict")
+        self.args.update(new_args)
+        for element in self.elements:
+            element.arguments(new_args)
+
+    ###########################################################################
+    # Math function                                                           #
+    ###########################################################################
+    def __add__(left, right):
+        if isinstance(left, QobjEvo):
+            self = left
+            other = right
+        else:
+            self = right
+            other = left
+        if not isinstance(other, (Qobj, QobjEvo, numbers.Number)):
+            return NotImplemented
+        res = self.copy()
+        res += other
+        return res
+
+    def __radd__(self, other):
+        if not isinstance(other, (Qobj, QobjEvo, numbers.Number)):
+            return NotImplemented
+        res = self.copy()
+        res += other
+        return res
```

Comment:
Yes, in cython's docs, to support both pre and post 3.0, they suggest have both `__op__(left, right)` and `__rop__(self, other)` or use `c_api_binop_methods=True`. I used to former since the `__rop__` methods already existed in `QobjEvo`. We can choose to use any of the method for Qutip. `c_api_binop_methods=True` may be safer since less code and tests are needed.

Reply:
Ok, sounds reasonable to me.  As it stands right now, I wouldn't at all be surprised if Cython 3.0 is a breaking change for us, but it's a good idea to start getting ready.